src/librustdoc/html/render/mod.rs

```diff
@@ -884,9 +884,11 @@ fn render_assoc_item(
                     ItemType::TyMethod
                 };
 
-                href(did.expect_real(), cx)
-                    .map(|p| format!("{}#{}.{}", p.0, ty, name))
-                    .unwrap_or_else(|| format!("#{}.{}", ty, name))
+                match href(did.expect_real(), cx) {
+                    Ok(p) => Some(format!("{}#{}.{}", p.0, ty, name)),
+                    Err(HrefError::DocumentationNotBuilt) => None,
```

Comment:
Ok, I think I see why I was confused - rustdoc can never be missing the documentation for the current crate, because it's building it right now.

Can you add an assert that `ty == ItemType::TyMethod` with a comment that TyMethods only appear for extern crates?

Reply:
This assert makes [this test](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/extern-default-method.rs) fail (with [this include](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/auxiliary/rustdoc-extern-default-method.rs)).

It seems it's possible to have a `ItemType::Method` when reexporting a struct with a method for which doc has not been built. I changed this PR to still allow this link as it's valid

Reply:
Do you happen to know why the test didn't fail without the assert? It seems to be testing that the `method.` link is generated - maybe it also needs to test that a `tymethod.` link is generated too?

Reply:
As we are importing and using the struct in the current crate, we are importing its impl with its method, but the source is in the dependency and the doc is not built there...

The existing assertion was not on the link itself, but on the presence of the function in the doc. I added assertions on the links generated